testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionConcurrencyTest.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.testsuite.model.session;
+
+import org.junit.Test;
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.models.UserSessionProvider;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+
+
+@RequireProvider(UserSessionProvider.class)
+public class UserSessionConcurrencyTest extends KeycloakModelTest {
+
+    private String realmId;
+    private static final int CLIENTS_COUNT = 10;
+
+    @Override
+    public void createEnvironment(KeycloakSession s) {
+        RealmModel realm = s.realms().createRealm("test");
+        realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
+        realm.setSsoSessionIdleTimeout(1800);
+        realm.setSsoSessionMaxLifespan(36000);
+        realm.setClientSessionIdleTimeout(500);
+        this.realmId = realm.getId();
+
+        s.users().addUser(realm, "user1").setEmail("user1@localhost");
+        s.users().addUser(realm, "user2").setEmail("user2@localhost");
+
+        for (int i = 0; i < CLIENTS_COUNT; i++) {
+            s.clients().addClient(realm, "client" + i);
+        }
+    }
+
+    @Override
+    public void cleanEnvironment(KeycloakSession s) {
+        s.realms().removeRealm(realmId);
+    }
+
+    @Override
+    protected boolean isUseSameKeycloakSessionFactoryForAllThreads() {
+        return true;
+    }
+
+    @Test
+    public void testConcurrentNotesChange() {
```

Comment:
What exactly is being tested here? When this test can fail and what determines a successful run?

Reply:
During development, I was getting some `ConcurrentModificationException` so I created this test to reproduce. This is doing a similar thing as `TokenManager` does [here](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L539-L564). 

Maybe we can add some assertions like there should be some amount of clients session and each has some value of the note. I will add it.